boto3/dynamodb/types.py

```diff
@@ -0,0 +1,287 @@
+            return True
+        elif isinstance(value, bytearray):
+            return True
+        elif six.PY3 and isinstance(value, six.binary_type):
+            return True
+        return False
+
+    def _is_set(self, value):
+        if isinstance(value, (set, frozenset)):
+            return True
+        return False
+
+    def _is_type_set(self, value, type_validator):
+        if self._is_set(value):
+            if False not in map(type_validator, value):
+                return True
+        return False
+
+    def _is_map(self, value):
+        if isinstance(value, Mapping):
+            return True
+        return False
+
+    def _is_list(self, value):
+        if isinstance(value, list):
+            return True
+        return False
+
+    def _serialize_null(self, value):
+        return True
+
+    def _serialize_bool(self, value):
+        return value
+
+    def _serialize_n(self, value):
+        number = str(DYNAMODB_CONTEXT.create_decimal(value))
+        if number in ['Infinity', 'NaN']:
+            raise TypeError('Infinity and NaN not supported')
+        return number
+
+    def _serialize_s(self, value):
+        return value
+
+    def _serialize_b(self, value):
+        if isinstance(value, Binary):
+            value = value.value
+        return value
+
+    def _serialize_ss(self, value):
+        return [self._serialize_s(s) for s in value]
+
+    def _serialize_ns(self, value):
+        return [self._serialize_n(n) for n in value]
+
+    def _serialize_bs(self, value):
+        return [self._serialize_b(b) for b in value]
+
+    def _serialize_l(self, value):
+        return [self.serialize(v) for v in value]
+
+    def _serialize_m(self, value):
+        return dict([(k, self.serialize(v)) for k, v in value.items()])
+
+
+class TypeDeserializer(object):
+    """This class deserializes DynamoDB types to Python types."""
+    def deserialize(self, value):
+        """The method to deserialize the DynamoDB data types.
+
+        :param value: A DynamoDB value to be deserialized to a pythonic value.
+            Here are the various conversions:
+
+            DynamoDB                                Python
+            --------                                ------
+            {'NULL': True}                          None
+            {'BOOL': True/False}                    True/False
+            {'N': str(value)}                       Decimal(str(value))
+            {'S': string}                           string
+            {'B': bytes}                            Binary(bytes)
+            {'NS': [str(value)]}                    set([Decimal(str(value))])
+            {'SS': [string]}                        set([string])
+            {'BS': [bytes]}                         set([bytes])
+            {'L': list}                             list
+            {'M': dict}                             dict
+
+        :returns: The pythonic value of the DynamoDB type.
+        """
+
+        if not value:
```

Comment:
Let's be strict on this. When I was porting this over, boto had some logic that just returned the attribute if ``len(value) > 1 or not value``. For some reason, I was thinking that it is possible to get an empty dictionary returned by dynamodb, but now that I think about it, it is not possible. If the value is empty, it will atleast have the type of the value.